### PR TITLE
Adding client retries

### PR DESCRIPTION
**Pull request by askpt**
Added the retry mechanism to the Orleans connection.
Screen Shot 2022-01-31 at 15 06 43

For the screenshot above, I set the logger to the `Warning` level in order to show the retries.

### CODE REVIEW

File: 5a/Client/src/Program.fs [hunk 1]
```diff
@@ -82,17 +85,20 @@ let main _args =
...
         (fun error ->
             task {
-                if not (error = null) then
-                    printfn $"Error connecting to cluster: ${error}"
-                    return false
-                else
-                    printfn "Connected to cluster"
-                    return true
+                Async.Sleep(TimeSpan.FromSeconds(1)) |> Async.RunSynchronously
+                attempts <- attempts + 1
```
Comment (PiotrJustyna) [inline 5a/Client/src/Program.fs:89]: Nah, no need for mutation here :) Let's try to use recursion instead.

File: 5a/Client/src/Program.fs [hunk 2]
```diff
@@ -82,17 +85,20 @@ let main _args =
     client.Connect
         (fun error ->
             task {
-                if not (error = null) then
-                    printfn $"Error connecting to cluster: ${error}"
-                    return false
-                else
-                    printfn "Connected to cluster"
-                    return true
+                Async.Sleep(TimeSpan.FromSeconds(1)) |> Async.RunSynchronously
```
Comment (PiotrJustyna) [inline 5a/Client/src/Program.fs:88]: can we not run it asynchronously using the `!` syntax?

File: 5a/Client/src/Program.fs [hunk 3]
```diff
@@ -82,17 +85,20 @@ let main _args =
...
+
+                return if attempts > maxAttempts then false
+                    else not(error = null)
             })
     |> Async.AwaitTask
     |> Async.RunSynchronously
 
-    let grain = client.GetGrain<IHelloWorld>(0)
+    // Generate random grain number key
+    let key = (Random().Next() % 100)
```
Comment (PiotrJustyna) [inline 5a/Client/src/Program.fs:99]: Do we need a random key? Not that it's wrong, just curious what you were testing.

**Review by PiotrJustyna — COMMENTED** [review]
Just minor comments, thank you 👍

File: 5a/Client/src/Program.fs (hunk 3 above)
Comment (askpt) [reply 5a/Client/src/Program.fs:99]: I made this change to be consistent with: https://github.com/PiotrJustyna/road-to-orleans/blob/003ec9a5b13d01fd4004b088620da41f892f97da/5/Client/src/HelloWorldClientHostedService.cs#L23

File: 5a/Client/src/Program.fs (hunk 2 above)
Comment (askpt) [reply 5a/Client/src/Program.fs]: https://github.com/PiotrJustyna/road-to-orleans/pull/312/commits/45ca9ff16c27150a2d489cc01c8761d137ac5969

File: 5a/Client/src/Program.fs (hunk 1 above)
Comment (askpt) [reply 5a/Client/src/Program.fs:89]: As discussed, the connect function needs a boolean to be returned in order to know if it should retry or not, which in this case we need to keep state and mutate the `attempts` variable.

Comment (PiotrJustyna) [reply 5a/Client/src/Program.fs:89]: You were 100% right, I stand corrected!

**Review by PiotrJustyna — APPROVED** [review]
Very nice, thank you 🎉

**Pull request merged**